testing/cli/test_cli_sct_compute_mtr.py

```diff
@@ -9,6 +11,33 @@
 @pytest.mark.sct_testing
 @pytest.mark.usefixtures("run_in_sct_testing_data_dir")
 def test_sct_compute_mtr_no_checks():
-    """Run the CLI script without checking results.
-    TODO: Check the results. (This test replaces the 'sct_testing' test, which did not implement any checks.)"""
+    """Run the CLI script without checking results."""
     sct_compute_mtr.main(argv=['-mt0', 'mt/mt0.nii.gz', '-mt1', 'mt/mt1.nii.gz'])
+
+    # Comparate ground truth mtr file to new generated one
+    ground_truth_mtr = Image('mt/mtr.nii.gz')
+    output_mtr = Image('mtr.nii.gz')
+    assert numpy.linalg.norm(ground_truth_mtr.data - output_mtr.data) <= 0.001
+
+
+@pytest.mark.sct_testing
+@pytest.mark.usefixtures("run_in_sct_testing_data_dir")
+def test_sct_compute_mtr_with_int16_image_type(tmp_path):
+    """Run the CLI script with int_16 image type"""
+
+    mt0_path = str(tmp_path / 'mt0_int16.nii.gz')
+    mt1_path = str(tmp_path / 'mt0_int16.nii.gz')
```

Comment:
> I also tried to add the check to see the difference between the two vectors but it seems that the content of the ouput mtr is only filled with 0:

I think the highlighted section above might be the cause! (Same filepath for both `mt0_path` and `mt1_path`, hehe.)

Reply:
... I will hide myself forever after that 🤦‍♂️ 

You were right, it was the only reason. Since the generation is good now I took the time to comparate files and it seems that the actual `sct_testing_data/mt/mtr.nii.gz` is not the same results from what I get by computing mtr from `sct_testing_data/mt/mt0.nii.gz` and `sct_testing_data/mt/mt1.nii.gz`.

I don't know if the problem is the function or the "ground truth" since the only test we had (ie: `test_sct_compute_mtr_no_checks`) did not check the results before.

@joshuacwnewton Do we have other way to get a mtr to create a ground truth ? (concurrent tool ? database ?)

Reply:
> Since the generation is good now I took the time to comparate files and it seems that the actual `sct_testing_data/mt/mtr.nii.gz` is not the same results from what I get by computing mtr from `sct_testing_data/mt/mt0.nii.gz` and `sct_testing_data/mt/mt1.nii.gz`.

Thank you for testing this! 

To be completely transparent, the dataset `sct_testing_data` (as well as many of the original tests) were created [6 years ago](https://github.com/spinalcordtoolbox/sct_testing_data/tree/master/mt), but I only joined the team 1.5 years ago. So, often I am finding things out just as you are! :p 

I was making an assumption when I suggested using `mtr.nii.gz`, but my assumption may have been wrong here. :sweat_smile: 

---

My first thought was that maybe the `mtr.nii.gz` file was actually generated by comparing `mt1.nii.gz` and `
mt0_reg_slicereg_goldstandard.nii.gz `, instead. Since, MTR is usually calculated between _coregistered_ MT0 and MT1 images (you can find more info in our tutorial [here](https://spinalcordtoolbox.com/user_section/tutorials/registration-to-template/mtr-computation.html)). 

If I make this change to your test, then we're able to get closer to matching MTR images: 

![Screenshot from 2022-01-12 14-02-04](https://user-images.githubusercontent.com/16181459/149204979-786ebc21-4659-42ba-87b6-2059bf986cf9.png)

While not identical, this could be good enough for us to write a comparison for now? e.g. within 0.1 rather than 0.001. 

(Still, I don't know the exact conditions the `mtr.nii.gz` file was generated in... :thinking:)

---

However, the other problem I'm noticing is that the MTR images contain both `nan` and `-inf` values. 

![Screenshot from 2022-01-12 14-11-25](https://user-images.githubusercontent.com/16181459/149206202-6531c7ca-4f03-4caa-9dc4-1e39f30ca385.png)

I wasn't expecting that! And it makes performing `numpy.linalg.norm(ground_truth_mtr.data - output_mtr.data)` a bit harder. For now, I think we can use [`numpy.isfinite`](https://stackoverflow.com/a/11620982) to just compare the finite values. (Though, the `nan`/`inf` behavior might require deeper investigation later on.)

When I make all of the changes above, I think I've got a working test. Would it be okay if I push the changes to your branch in a separate commit? :)

Reply:
> (Still, I don't know the exact conditions the mtr.nii.gz file was generated in... 🤔)

IIRC it was using the batch_processing, so yes: after co-registering the data together. So if the goal is to assess this numerical results, i would personally not use the existing mtr data as ground truth

Reply:
> So if the goal is to assess this numerical results, i would personally not use the existing mtr data as ground truth

Apologies, but I'm not sure I understand...

My expectation was that we would want to make sure the MTR results are roughly the same as what they were 6 years ago. Wouldn't comparing the current MTR output to the previous output be a worthwhile test?

Or are you suggesting that it would be better to try and find some objective measure of "correct" results to use as ground truth to compare against instead? (This seems much more difficult given our present testing data... Hmm.)

Reply:
sorry for my lack of clarity: i meant that we should compare an mtr that is computed from the same input data (the current mtr is not computed from the same input data, because IIRC one has been coregistered and resampled). So one solution would be to compute another mtr, from the native data (unregistered) and add it to the testing data (eg could be called mtr_not_registered.nii.gz)

Reply:
@joshuacwnewton  Yes feel free to add commits, I'm not conservative on my PR ;)

@jcohenadad But if we compute a new mtr from native data (uneregistered) we will not be sure that the mtr is good since we want to test our function. Will you be able to do an exceptional manual validation to make it a ground truth ? Or do you propose to compute another mtr but with an other tool ?

Reply:
> @jcohenadad But if we compute a new mtr from native data (uneregistered) we will not be sure that the mtr is good since we want to test our function. Will you be able to do an exceptional manual validation to make it a ground truth ? Or do you propose to compute another mtr but with an other tool ?

right, chicken-and-egg problem 😉   the mtr we currently consider 'ground truth' isn't necessarily better than the one i am proposing to create because it would be using the same code (ie: the code is so trivial that we trust it... by maybe we shouldn't?). Alternatively you could create a synthetic NIfTI mt0 and mt1 and use that for the functional test, eg we do that for some tests, eg [here](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/6ca81d72c9e3eef265ba3a0ff63d8297b4372f3d/spinalcordtoolbox/testing/create_test_data.py#L120) and [here](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/6ca81d72c9e3eef265ba3a0ff63d8297b4372f3d/testing/api/test_resampling.py#L21).

Reply:
@jcohenadad Even if we use synthetic NifTI image for mt0 and mt1, the mtr groundtruth will always depend of the same code we want to test.

I don't have the medical knowledge, but if you say that there is **no other way** to compute an MTR that we can use as a ground truth I think it's OK to generate a new MTR based on float32 image as you did before (no reason to think it does not work lately) and store it as a ground truth. We can just add some comments in the test to explain the situation so future contributor can take it in account if one day they found the ground truth to be no so ground.

I just need a greenlight from you @jcohenadad since it's more a data research question than a technical software issue.

Reply:
> @jcohenadad Even if we use synthetic NifTI image for mt0 and mt1, the mtr groundtruth will always depend of the same code we want to test.

given that mtr is based on a [simple equation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/6ca81d72c9e3eef265ba3a0ff63d8297b4372f3d/spinalcordtoolbox/qmri/mt.py#L50) we could something like: mt0=100, mt1=70, "assess mtr = 100*(100-70)/100 = 30"

Reply:
@jcohenadad Oh! I didn't saw it was that simple 😄 

I will try to create an `mt0` `mt1` and `mtr` and compare it with the system output. Will be a good way for me to rampup on NifTI image data and we will have a ground truth for the future!